modules/core/src/main/scala/com.snowplowanalytics.iglu/client/resolver/Resolver.scala

```diff
@@ -181,6 +187,82 @@ final case class Resolver[F[_]](repos: List[Registry], cache: Option[ResolverCac
+      allIgluCentral.contains(repo)
+    }
+    (igluCentral.isEmpty || igluCentral.values.exists(
+      _.errors.exists(_ == RegistryError.NotFound)
+    )) && custom.values.flatMap(_.errors).forall(_ == RegistryError.NotFound)
+  }
+
+  /**
+   * Looks up all the schemas with the same model until `maxSchemaKey`.
+   * For the schemas of previous revisions, it starts with addition = 0
+   * and increments it until a NotFound.
+   *
+   * @param maxSchemaKey The SchemaKey until which schemas of the same model should get returned
+   * @return All the schemas if all went well, [[Resolver.SchemaResolutionError]] with the first error that happened
+   *         while looking up the schemas if something went wrong.
+   */
+  def lookupSchemasUntil(
+    maxSchemaKey: SchemaKey
+  )(implicit
+    F: Monad[F],
+    L: RegistryLookup[F],
+    C: Clock[F]
+  ): F[Either[SchemaResolutionError, NonEmptyList[SelfDescribingSchema[Json]]]] = {
+    def go(
+      current: SchemaVer.Full,
+      acc: NonEmptyList[SelfDescribingSchema[Json]]
+    ): F[Either[SchemaResolutionError, NonEmptyList[SelfDescribingSchema[Json]]]] = {
+      val currentSchemaKey = maxSchemaKey.copy(version = current)
+      lookupSchema(currentSchemaKey).flatMap {
+        case Left(e) =>
+          if (current.addition === 0)
+            Monad[F].pure(Left(SchemaResolutionError(currentSchemaKey, e)))
+          else if (current.revision < maxSchemaKey.version.revision && isNotFound(e))
+            go(current.copy(revision = current.revision + 1, addition = 0), acc)
+          else
+            Monad[F].pure(Left(SchemaResolutionError(currentSchemaKey, e)))
+        case Right(json) =>
+          if (current.revision < maxSchemaKey.version.revision)
+            go(
+              current.copy(addition = current.addition + 1),
+              SelfDescribingSchema(SchemaMap(currentSchemaKey), json) :: acc
+            )
+          else if (current.addition < maxSchemaKey.version.addition)
+            go(
+              current.copy(addition = current.addition + 1),
+              SelfDescribingSchema(SchemaMap(currentSchemaKey), json) :: acc
+            )
+          else
+            Monad[F].pure(
+              Right((SelfDescribingSchema(SchemaMap(currentSchemaKey), json) :: acc).reverse)
+            )
+      }
+    }
+
+    val firstSchemaKey =
+      maxSchemaKey.copy(version = SchemaVer.Full(maxSchemaKey.version.model, 0, 0))
+
+    EitherT(lookupSchema(firstSchemaKey))
+      .leftMap(SchemaResolutionError(firstSchemaKey, _))
+      .flatMap { firstJson =>
+        val acc = NonEmptyList.one(SelfDescribingSchema(SchemaMap(firstSchemaKey), firstJson))
+
+        if (maxSchemaKey.version.addition == 0 && maxSchemaKey.version.revision == 0)
+          EitherT.rightT[F, SchemaResolutionError](acc)
+        else
+          EitherT(go(SchemaVer.Full(maxSchemaKey.version.model, 0, 1), acc))
+      }
+      .value
```

Comment:
I tried with a for loop but I thought that it looks ~better~ less worse like that

Reply:
How about doing it like [this](https://github.com/snowplow/iglu-scala-client/commit/a6a8616c5f216f0367be3f673159769f4fb2b9a9)?

Reply:
I never considered that `go` could have a different return type than the function :facepalm: